### PR TITLE
GTM logic refactored

### PR DESCRIPTION
**Pull request by maxsam4**
GTM logic has been simplified and made more flexible.
Some work will be required by the dApp team to incorporate the changes.

### CODE REVIEW

File: contracts/proxy/GeneralTransferManagerProxy.sol [hunk 1]
```diff
@@ -26,6 +26,7 @@ contract GeneralTransferManagerProxy is GeneralTransferManagerStorage, ModuleSto
     {
         require(_implementation != address(0), "Implementation address should not be 0x");
         _upgradeTo(_version, _implementation);
+        _setDefaults();
```
Comment (adamdossa) [inline contracts/proxy/GeneralTransferManagerProxy.sol]: Wonder whether this should be in the GTM initialisation function instead? Possibly in a future upgraded GTM this function may change.

Comment (adamdossa) [reply contracts/proxy/GeneralTransferManagerProxy.sol]: I guess it only gets called the first time the proxy is created and the logic contract is set (not on future upgradeToAndCall calls) so this is fine.

File: contracts/modules/TransferManager/GTM/GeneralTransferManager.sol [hunk 2]
```diff
@@ -150,67 +107,133 @@ contract GeneralTransferManager is GeneralTransferManagerStorage, TransferManage
...
+    /**
+    * @notice Modifies the successful checks required for transfers.
+    * @param _transferType Type of transfer (0 = General, 1 = Issuance, 2 = Redemption)
+    * @param _fromValidKYC Defines if KYC is required for the sender
+    * @param _toValidKYC Defines if KYC is required for the receiver
+    * @param _fromRestricted Defines if transfer time restriction is checked for the sender
+    * @param _toRestricted Defines if transfer time restriction is checked for the receiver
+    */
+    function modifyTransferRequirementsMulti(
+        uint256[] memory _transferType,
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/GTM/GeneralTransferManager.sol]: a plural form should be used here

File: contracts/modules/TransferManager/GTM/GeneralTransferManager.sol [hunk 3]
```diff
@@ -150,67 +107,133 @@ contract GeneralTransferManager is GeneralTransferManagerStorage, TransferManage
...
+    }
+
+    function _modifyTransferRequirements(
+        uint256 _transferType,
+        bool _fromValidKYC,
+        bool _toValidKYC,
+        bool _fromRestricted,
+        bool _toRestricted
+    ) internal {
+        transferRequirements[_transferType] =
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/GTM/GeneralTransferManager.sol:221]: can add a sanity check on `transferType`

**Review by satyamakgec — COMMENTED** [review]
The approach is good on the contract side but it will little bit harden the dApp work. dApp team needs to create some scenarios (combination of true or false) when issuer wants to change in the transfers flag.
And redemption/ allowAllWhitelistTransfers will little bit costlier.

**Comment by maxsam4** [comment]
@satyamakgec `allowAllWhitelistTransfers` will actually be cheaper. Ones less SLOAD is required in the new logic. 

Only `allowAllTransfers` and `allowAllBurnTransfers` will be slightly more expensive but I don't think anyone will use them in practice.

**Review by satyamakgec — APPROVED** [review]

**Review by adamdossa — APPROVED** [review]

**Pull request merged**